### PR TITLE
Avoid single_path_optimization if offsets are non-zero

### LINKED ISSUE #16461: Hexbin if singular and mincnt used
### Bug report

**Bug summary**

Related to #3038, #2863 and [this stackoverflow bug](https://stackoverflow.com/questions/14780362/is-it-possible-to-plot-a-list-of-identical-points-on-a-matplotlib-hexbin-plot): if only identical (or one single) points are given, 

This seems to *still* be a bug when using `mincnt > 0`

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt

mytens = [np.array([[-1, 1], [1, 1]]),
          np.array([[-1, 1], [-1, -1], [-1, -1]]),
          np.array([[-1, 1]])]
fig, axs = plt.subplots(nrows=2, ncols=2, sharex="all", sharey="all", squeeze=False)
axs[0, 0].hexbin(
    x=mytens[0][:, 0],
    y=mytens[0][:, 1],
    mincnt=1,
    gridsize=4
)
axs[0, 0].set_title("correct point 1")
axs[0, 1].hexbin(
    x=mytens[1][:, 0],
    y=mytens[1][:, 1],
    mincnt=1,
    gridsize=4
)
axs[0, 1].set_title("correct point 2")
axs[1, 0].hexbin(
    x=mytens[2][:, 0],
    y=mytens[2][:, 1],
    mincnt=1,
    gridsize=4
)
axs[1, 0].set_title("actual point not plotted, instead at 0,0!")
axs[1, 1].hexbin(
    x=mytens[2][:, 0],
    y=mytens[2][:, 1],
    # mincnt=0,  # if not given, 0 is assigned internally
    gridsize=4
)
axs[1, 1].set_title("actual point plotted, expected behaviour")
plt.show()



```

**Actual outcome**

removeme

**Expected outcome**

Point on the bottom left should be on the left top. The plotted point (0,0) does not even exist in the given data.

**Matplotlib version**

  * Operating system: Win 10
  * Matplotlib version: 3.1.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version:  3.7.6 
  * Jupyter version (if applicable): - 
  * Other libraries: 
    - numpy 1.18.1  
    - (But same problem with torch)

### PR DESCRIPTION
**Pull request by dstansby**
This is a stopgap to fix #16461 - I am guessing this is made redundant by https://github.com/matplotlib/matplotlib/pull/13696, but is a simpler fix and probably easier to get into 3.2.0.

### CODE REVIEW

**Comment by anntzer** [comment]
Perhaps do a test like given in https://github.com/matplotlib/matplotlib/issues/16461 but settings x/ylim so that in the correct case the entire patch is out of bounds, but would be drawn at (0, 0) otherwise, and then check_figures_equal() against an empty plot?

**Comment by dstansby** [comment]
That worked well - the test I've added fails on current master but passes here.

**Comment by anntzer** [comment]
If we need to change baseline images (especially non-hexbin-related ones) I'd really rather just go with https://github.com/matplotlib/matplotlib/pull/13696 -- this bug has apparently been present for a *long* time, I'm not convinced it's RC to have it fixed in 3.2.  Or if it helps I can make VectorTransform public in #16396; I don't like it but it's not the end of the world.

**Comment by dstansby** [comment]
Agree with not needing to put this in 3.2. Won't the baseline images have to change anyway at some point though?

**Comment by anntzer** [comment]
I don't think so.  I think (not 100% sure?) here you disabled single_path_optimization when offsets are nonzero, but they really only need to be disabled when offsets are nonzero *and* offset_position="data".  When offsets are nonzero and offset_position="screen" (which will always be the case after #13696), the old baseline images are still fine.

**Comment by dstansby** [comment]
Ah yes, true - I can check for offset_position = 'data' here and not worry about the baseline images then.

Surely going through either of the two code paths should give identical results, but it doesn't look like it does because the images have changed. I guess the question is for offset_position = 'screen', which code path is doing the right thing?

**Comment by anntzer** [comment]
Basically single_path_optimization is slightly "wrong", because it "rounds" the markers to the closest pixel (it's fundamentally the original reason I wrote mplcairo...), see e.g. #7233.
But there are cases, e.g. ticks, where the rounding is actually useful because if you try to draw short ticks antialiased "between" two rows of pixel, they look particularly bad.

File: lib/matplotlib/collections.py [hunk 1]
```diff
@@ -345,6 +345,14 @@ def draw(self, renderer):
                     and extents.height < self.figure.bbox.height):
                 do_single_path_optimization = True
 
+        # The single path optimization code path is currently broken when
+        # offsets are specified
+        # (https://github.com/matplotlib/matplotlib/issues/16496)
+        # This should be fixed eventually, but for now don't use the optimized
+        # code path if some of the offsets are non-zero
+        if not np.all(offsets == 0) and self._offset_position == 'data':
```
Comment (timhoffm) [inline lib/matplotlib/collections.py]:
```suggestion
        if np.any(offsets != 0) and self._offset_position == 'data':
```
I would find this slightly more readable but won't block over this.

**Comment by QuLogic** [comment]
So this is replaced by #13696?

**Comment by anntzer** [comment]
I think so? feel free to reopen if not.

**Pull request closed without merging**